src/aspire/abinitio/commonline_c3_c4.py

```diff
@@ -231,21 +237,270 @@ def _estimate_inplane_rotations(
     # Secondary Methods for computing outer product #
     #################################################
 
-    def _self_clmatrix_c3_c4(self, pf, order, max_shift, shift_step):
-        # return sclmatrix
-        pass
+    def _self_clmatrix_c3_c4(self):
+        """
+        Find the single pair of self-common-lines in each image assuming that the underlying
+        symmetry is C3 or C4.
+        """
+        pf = self.pf.copy()
+        n_img = self.n_img
+        n_theta = self.n_theta
+        max_shift_1d = np.ceil(2 * np.sqrt(2) * self.max_shift)
```

Comment:
The comment in the matlab code is that "the maximum shift occurs at a diagonal direction." So 2√2 would be taking the diagonal and doubling it (presumably to account for the length of the diagonal on both sides of the origin). I'm actually a little confused by this because when it is used in `_generate_shift_phase_and_filter()`, the number of "shifts to try" is calculated as `n_shifts = int(np.ceil(2 * max_shift / shift_step + 1))`, which seems to double the number of shifts again. Maybe this should just be √2.

Reply:
So do I understand this correctly: `max_shift` describes the half-side length of a square in which we search for a shift? In other words, the search space is [–`max_shift`, `max_shift`]². That would mean that the maximum _difference_ of shifts is 2√2 times `max_shift`. But why is that the relevant distance here?

Reply:
I'm not sure that distance _is_ relevant. My best guess is that it was a mistake in the original matlab implementation. Which can be seen [here](https://github.com/PrincetonUniversity/aspire/blob/e697eee5bb185ab1962f1d6e5becaf9caa277732/abinitio/Cn/calc_shift_phases.m) on line 24. 

This same method for finding shift phases is used for the common line sync method and there is no multiplication by 2√2.  I think this should be taken out.

Reply:
Took this out for now.

Reply:
Ok. And tests still pass?

Reply:
Yes, but we are using a simulation with zero offsets. Maybe a test should be added with shifts.